File: src/main/scala/mrtjp/fengine/assemble/TileMapPathFinder.scala
package mrtjp.fengine.assemble

import mrtjp.fengine.TileCoord
import mrtjp.fengine.tiles.{FETile, FETileMap}

import scala.collection.mutable

class TileMapPathFinder(map:FETileMap, coord:TileCoord, propFunc:(Int, Int) => Boolean)
{
    private val open = mutable.Queue[PathFinderNode]()
    private val openSet = mutable.HashSet[PathFinderNode]()
    private val closedSet = mutable.HashSet[PathFinderNode]()

    val portToOutputRegisters:mutable.HashMap[Int, mutable.HashSet[Int]] = mutable.HashMap[Int, mutable.HashSet[Int]]()
    val portToInputRegisters:mutable.HashMap[Int, mutable.HashSet[Int]] = mutable.HashMap[Int, mutable.HashSet[Int]]()

    openInitial(coord, propFunc)

    private def openInitial(coord:TileCoord, propFunc:(Int, Int) => Boolean):Unit =
    {
        openNext(coord, propFunc, PathFinderNode.apply)
    }
    private def openFrom(prev:PathFinderNode, coord:TileCoord, propFunc:(Int, Int) => Boolean):Unit =
    {
        openNext(coord, propFunc, prev.moveTo)
    }
    private def openNext(coord:TileCoord, propFunc:(Int, Int) => Boolean, nodeFactory:(TileCoord, Int, Int) => PathFinderNode):Unit =
    {
        for (s <- 0 until 6) for (p <- 0 until 16) if (propFunc(s, p)) {
            val move = nodeFactory(coord, s, p)
            if (!openSet.contains(move) && !closedSet.contains(move)) {
                open += move
                openSet += move
            }
        }
    }

    private def collect(root:PathFinderNode, tile:FETile, coord:TileCoord, dir:Int, port:Int):Unit =
    {
        val initialPort = root.port // Initial port at start of pathfinding which lead to discovery of this new tile

        tile.getOutputRegister(dir, port) match {
            case Some(outputRegID) => // Pathfinding lead to tile that has output register
                val inputRegisters = portToInputRegisters.getOrElseUpdate(initialPort, new mutable.HashSet[Int]())
                inputRegisters += outputRegID // Another tile's output is root tile's input
            case None =>
        }
        tile.getInputRegister(dir, port) match {
            case Some(inputRegID) =>
                val outputRegisters = portToOutputRegisters.getOrElseUpdate(initialPort, new mutable.HashSet[Int]())
                outputRegisters += inputRegID // Another tile's input is root tile's output
            case None =>
        }
    }

    private def iterate():Unit =
    {
        if (open.isEmpty) return

        val next = open.dequeue()
        openSet.remove(next)

        val newPos = next.pos.offset(next.dir)
        map.getTile(newPos) match {
            case Some(tile) =>
                val fromDir = TileCoord.oppositeDir(next.dir)
                val pfunc = tile.propagationFunc(fromDir, next.port)
                collect(next.rootNode, tile, newPos, fromDir, next.port)
                openFrom(next.rootNode, newPos, pfunc)

                closedSet += next
            case None =>
        }
    }

    def step():Unit =
    {
        iterate()
    }

    def isFinished:Boolean = open.isEmpty

    def result():PathFinderResult = new PathFinderResult(
        portToOutputRegisters.map(kv => (kv._1, kv._2.toSet)).toMap,
        portToInputRegisters.map(kv => (kv._1, kv._2.toSet)).toMap)
}

File: src/main/java/mrtjp/fengine/assemble/TileMapPathFinder.java
package mrtjp.fengine.assemble;

import mrtjp.fengine.TileCoord;
import mrtjp.fengine.api.PropagationFunction;
import mrtjp.fengine.tiles.FETile;
import mrtjp.fengine.tiles.FETileMap;

import java.util.*;
import java.util.stream.Collectors;

public class TileMapPathFinder { //TODO implement Stepper

    private final Queue<PathFinderNode> open = new LinkedList<>();
    private final HashSet<PathFinderNode> openSet = new HashSet<>();
    private final HashSet<PathFinderNode> closedSet = new HashSet<>();

    private final HashMap<Integer, HashSet<Integer>> portToOutputRegisters = new HashMap<>();
    private final HashMap<Integer, HashSet<Integer>> portToInputRegisters = new HashMap<>();

    private final PathFinderManifest manifest;
    private final FETileMap map;

    public TileMapPathFinder(PathFinderManifest manifest, FETileMap map, TileCoord start, PropagationFunction startPropFunc) {
        this.manifest = manifest;
        this.map = map;
        openInitial(start, startPropFunc);
    }

    private void openInitial(TileCoord coord, PropagationFunction propFunc) {

        traverse(propFunc, (s, p) -> {
            TileCoord nextPos = coord.offset(s);
            int inputDir = TileCoord.oppositeDir(s);
            PathFinderNode opened = new PathFinderNode(nextPos, inputDir, p);
            if (!openSet.contains(opened) && !closedSet.contains(opened)) {
                open.add(opened);
                openSet.add(opened);
            }
        });
    }

    private void openNext(PathFinderNode prev) {

        FETile tile = map.getTile(prev.pos).orElse(null);
        if (tile == null) return;

        collect(prev, tile);

        PropagationFunction propFunc = tile.propagationFunc(prev.inputDir, prev.inputPort);
        traverse(propFunc, (s, p) -> {
            TileCoord nextPos = prev.pos.offset(s);
            int inputDir = TileCoord.oppositeDir(s);
            PathFinderNode opened = prev.moveTo(nextPos, inputDir, p);
            if (!openSet.contains(opened) && !closedSet.contains(opened)) {
                open.add(opened);
                openSet.add(opened);
            }
        });
    }

    private void collect(PathFinderNode prev, FETile tile) {

        int initialPort = prev.getRootNode().inputPort;
        int dir = prev.inputDir;
        int port = prev.inputPort;

        tile.getOutputRegister(dir, port).ifPresent(outputRegID -> {
            portToInputRegisters.computeIfAbsent(initialPort, i -> new HashSet<>()).add(outputRegID);
            manifest.addInputPropagationTrail(outputRegID, prev);
        });

        tile.getInputRegister(dir, port).ifPresent(inputRegID -> {
            portToOutputRegisters.computeIfAbsent(initialPort, i -> new HashSet<>()).add(inputRegID);
            manifest.addOutputPropagationTrail(inputRegID, prev);
        });
    }

    private void traverse(PropagationFunction propFunc, TraverseFunc traverseFunc) {
        for (int s = 0; s < 6; s++) {
            for (int p = 0; p < 16; p++) {
                if (propFunc.canPropagate(s, p)) traverseFunc.traverse(s, p);
            }
        }
    }

    public void step() {
        if (open.isEmpty()) return;

        // Get next node on queue
        PathFinderNode next = open.poll();
        openSet.remove(next);

        // Operate on it and queue additional nodes
        openNext(next);

        // Close this node
        closedSet.add(next);
    }

    public boolean isFinished() {
        return open.isEmpty();
    }

    public PathFinderResult result() {

        List<Integer> outputs = portToOutputRegisters.values().stream()
                .flatMap(Set::stream)
                .distinct()
                .sorted()
                .collect(Collectors.toList());
        List<Integer> inputs = portToInputRegisters.values().stream()
                .flatMap(Set::stream)
                .distinct()
                .sorted()
                .collect(Collectors.toList());

        return new PathFinderResult(outputs, inputs);
    }

    private interface TraverseFunc {

        void traverse(int dir, int port);
    }
}
